virtio-devices/src/balloon.rs

```diff
@@ -168,56 +177,85 @@ impl BalloonEpollHandler {
             used_count += 1;
 
-            let data_chunk_size = size_of::<u32>();
-
-            // The head contains the request type which MUST be readable.
-            if avail_desc.is_write_only() {
-                error!("The head contains the request type is not right");
-                return Err(Error::UnexpectedWriteOnlyDescriptor);
-            }
-            if avail_desc.len as usize % data_chunk_size != 0 {
-                error!("the request size {} is not right", avail_desc.len);
-                return Err(Error::InvalidRequest);
-            }
+            if ev_type == REPORTING_QUEUE_EVENT {
+                loop {
+                    if let Ok(hva) = mem.get_host_address(avail_desc.addr) {
+                        let res = unsafe {
+                            libc::madvise(
+                                hva as *mut libc::c_void,
+                                avail_desc.len as libc::size_t,
+                                libc::MADV_DONTNEED,
+                            )
+                        };
+                        if res != 0 {
+                            return Err(Error::MadviseFail(io::Error::last_os_error()));
+                        }
+                    } else {
+                        error!("Address {:?} is not available", avail_desc.addr);
+                        return Err(Error::InvalidRequest);
+                    }
 
-            let mut offset = 0u64;
-            while offset < avail_desc.len as u64 {
-                let addr = avail_desc.addr.checked_add(offset).unwrap();
-                let pfn: u32 = mem.read_obj(addr).map_err(Error::GuestMemory)?;
-                offset += data_chunk_size as u64;
-
-                let gpa = (pfn as u64) << VIRTIO_BALLOON_PFN_SHIFT;
-                if let Ok(hva) = mem.get_host_address(GuestAddress(gpa)) {
-                    let advice = match ev_type {
-                        INFLATE_QUEUE_EVENT => libc::MADV_DONTNEED,
-                        DEFLATE_QUEUE_EVENT => libc::MADV_WILLNEED,
-                        _ => return Err(Error::ProcessQueueWrongEvType(ev_type)),
-                    };
-                    // Need unsafe to do syscall madvise
-                    let res = unsafe {
-                        libc::madvise(
-                            hva as *mut libc::c_void,
-                            (1 << VIRTIO_BALLOON_PFN_SHIFT) as libc::size_t,
-                            advice,
-                        )
-                    };
-                    if res != 0 {
-                        return Err(Error::MadviseFail(io::Error::last_os_error()));
+                    if !avail_desc.has_next() {
+                        break;
                     }
-                } else {
-                    error!("Address 0x{:x} is not available", gpa);
+                    avail_desc = avail_desc
+                        .next_descriptor()
+                        .ok_or(Error::DescriptorChainTooShort)?;
```

Comment:
Instead of checking if the `avail_desc` has a next descriptor first and then pull this next descriptor, you can simplify like this:
```rust
if let Some(desc) = avail_desc.next_descriptor() {
        avail_desc = desc;
} else {
        break;
}
```

Reply:
Fixed.